core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -65,12 +65,18 @@ public static boolean ancestorOf(Table table, long snapshotId, long ancestorSnap
   }
 
   /**
-   * Traverses the history of the table's current snapshot and finds the oldest Snapshot.
-   * @return null if there is no current snapshot in the table, else the oldest Snapshot.
+   * Traverses the history of the table's current snapshot
+   * and finds the oldest Snapshot after or equal to the timestamp in milliseconds.
+   * @return null if there is no current snapshot in the table,
+   * else the oldest Snapshot after or equal to the timestamp in milliseconds.
    */
-  public static Snapshot oldestSnapshot(Table table) {
+  public static Snapshot oldestSnapshot(Table table, long timestampMillis) {
```

Comment:
I don't think the original purpose of this method and the timestamp lookup should be mixed together. This is no longer the "oldest snapshot" and is now the oldest snapshot newer than some timestamp. I think this should be a new method called `oldestAncestorAfter`.

Also, we need to consider boundary conditions. When resuming from a snapshot, if the previous snapshot is not found in the table's known snapshots, then we can't resume because there may have been unknown changes. This is similar. I think we should only return a snapshot from the `oldestAncestorAfter` method if its parent is still known and older than the timestamp. That's the only way we can ensure that there wasn't a snapshot after the timestamp that was aged off.

Reply:
I agree oldestAncestorAfter is a better name.
But I don't think NPE will be happen, the method indeed will only return a snapshot if its parent is still known and older than the timestamp.
For the “if the parent is no longer tracked” I am not sure when it will happen, based on my understanding, we does have safety check:
`while (current.parentId() != null && table.snapshot(current.parentId()).timestampMillis() >= timestampMillis)`
I think this logic will handle the parent is not existed case

Reply:
To be clear, I'm asking for 2 things here:
1. This should leave `oldestSnapshot` unchanged and should create a new method, `oldestAncestorAfter`.
2. This needs to avoid the `NullPointerException` that will be thrown when `table.snapshot(current.parentId())` returns `null` because the parent snapshot is no longer in table metadata. You can simulate this case by creating a couple snapshots and then expiring the first and asking for a timestamp older than the earliest snapshot in the table, but after the snapshot you expired.

Since there is some confusion here, I think this method also warrants some test cases to validate its behavior.

Reply:
@SreeramGarlapati and @zhengzengprc, can you address this?

Reply:
Thanks @rdblue, it is a very interesting problem to handle the expired snapshot. I can reproduce the problem in unit test. I am thinking about it is better way to loop the snapshots without the parentId
`  public static Snapshot snapshotAfter(Table table, long snapshotId) {
    Preconditions.checkArgument(table.snapshot(snapshotId) != null, "Cannot find parent snapshot: %s", snapshotId);

    Snapshot current = table.currentSnapshot();
    while (current != null) {
      if (current.parentId() == snapshotId) {
        return current;
      }

      current = table.snapshot(current.parentId());
    }

    throw new IllegalStateException(
        String.format("Cannot find snapshot after %s: not an ancestor of table's current snapshot", snapshotId));
  }`
It seems like we have similar issue with the above code block when some of the previous snapshot expired

Reply:
1. Leave oldestSnapshot unchanged and created new function oldestAncestorAfter
2. Handle the NPE in oldestAncestorAfter, and created unit test expired the first snapshot in the chain, and read stream from the timestamp of the expired snapshot

Reply:
To simplify the work here, I updated how `SnapshotUtil` works in #3259. That adds a way to create an `Iterable<Snapshot>` of the ancestors given some starting snapshot, `SnapshotUtil.ancestorsOf`. That should make implementing `oldestAncestorAfter` cleaner:

```java
  /**
   * Traverses the history of the table's current snapshot and finds the first snapshot after the given timestamp.
   *
   * @param table a table
   * @param timestampMillis a timestamp in milliseconds
   * @return the first snapshot after the given timestamp, or null if the current snapshot is older than the timestamp
   * @throws IllegalStateException if the first ancestor after the given time can't be determined
   */
  public static Snapshot oldestAncestorAfter(Table table, long timestampMillis) {
    if (table.currentSnapshot() == null) {
      // there are no snapshots or ancestors
      return null;
    }

    Snapshot lastSnapshot = null;
    for (Snapshot snapshot : currentAncestors(table)) {
      if (snapshot.timestampMillis() <= timestampMillis) {
        return lastSnapshot;
      }

      lastSnapshot = snapshot;
    }

    if (lastSnapshot != null && lastSnapshot.parentId() == null) {
      // this is the first snapshot in the table, return it
      return lastSnapshot;
    }

    throw new IllegalStateException(
        "Cannot find snapshot older than " + DateTimeUtil.formatTimestampMillis(timestampMillis));
  }
```

Reply:
Thanks @rdblue , that's really good work with the iterator approach to look up the ancestor. I just merge the latest master and use the api re-implement the function. Please take a look when you have time. Thanks